Optimize CellId

gassmoeller opened the pull request:
This PR allows to serialize a CellId object into a binary representation of the data instead of a string. I introduced this mostly because the current string serialization/deserialization is extremely slow. To illustrate: A function in ASPECT that serializes particle and CellId data, ships it via MPI and deserializes it afterwards is completely dominated by the serialization/deserialization of the CellId data (so even MPI transfer is fast compared to this code). Using the new binary representation the whole process takes 80% less time (and the CellId part still uses 40% of this remaining time).
I have a few questions before finishing this PR though:
- CellId currently does not know about the space dimension, but the binary format relies on it. I can make `to_binary` a template function, but I can not generate a template constructor, because for that the template needs to be deducible, which it is not. For now I went the ugly way and made it a constructor argument. Is there a nicer way? Should CellId for example be a template class? 
- Would it make sense to change the internal storage to a 64 bit unsigned int as suggested in the TODO comment of the class? This will increase the speed substantially, but it would limit the maximum number of levels in 3D to 21, and in 2D to 32, is this acceptable? Are there currently limits on the number of levels?
- Is there a reason `CellAccessor<dim,spacedim>::id()` is defined in the header `include/deal.II/grid/tria_accessor.h`? This is a bit annoying for development, because every change requires a full compile cycle for deal.II, and I do not see the advantage.

When we have settled on a way to go I will also add a test that checks the new functionality (it works already for me).

Rombur commented on source/grid/cell_id.cc:
```diff
@@ -21,13 +21,110 @@
+  for (; i<sizeof(unsigned int); ++i)
+    {
+      const unsigned int temp = static_cast<unsigned int> (binary_representation[i]);
+      coarse_cell_id = coarse_cell_id | (temp<<(i*8));
```
Is it always 8 or can the compiler decides to use a different (larger?) number of bits?

bangerth replied:
The C standard does not actually say. There used to be machines with 9 bits per byte but since everyone uses 8 bytes these days, I think it is safe to say that nobody in their right mind would propose a processor these days that does not follow this precedent. I think this is safe.

Rombur commented on source/grid/cell_id.cc:
```diff
@@ -21,13 +21,110 @@
+        }
+    }
+}
+
```
Three blank lines.

Rombur commented on source/grid/cell_id.cc:
```diff
@@ -21,13 +21,110 @@
+    {
+      for (unsigned int j=0; j<children_per_char; ++j)
+        {
+          child_indices[child_level] = (binary_representation[i] >> (dim*j)) & mask;
```
Can you explain what you are doing here.

Rombur commented on source/grid/cell_id.cc:
```diff
@@ -21,13 +21,110 @@
+        {
+          const unsigned int a = static_cast<unsigned int>(child_indices[child_level]);
+          const unsigned int shifted_child_index = a << (j*dim);
+          binary_representation[i] = binary_representation[i] | shifted_child_index;
```
What is `shifted_child_index`?

Rombur commented on source/grid/cell_id.cc:
```diff
@@ -21,13 +21,110 @@
+
+  return binary_representation;
+}
+
```
Three blank lines.

bangerth commented:
The data type should probably have been a template given that it makes no sense to compare `CellId` objects for 2d and 3d in one statement. I don't know whether we still want to make this sort of change. @tjhei ?

`CellAccessor::id` -- if it's cheap, it makes sense in the header file. If it is not, then it may as well move into the `.cc` file.

Rombur commented:
> Would it make sense to change the internal storage to a 64 bit unsigned int as suggested in the TODO comment of the class? This will increase the speed substantially, but it would limit the maximum number of levels in 3D to 21, and in 2D to 32, is this acceptable? Are there currently limits on the number of levels?

There is a limit on the number of levels because of p4est. I think that p4est can handle up to 11 levels per tree but one mesh can be compose of an arbitrary number of trees so the coarse cell id is arbitrary. So I think if we use a 64 bit unsigned int for `child_indices` and another oner `coarse_cell_id`, we should be good.

tjhei commented:
> CellId currently does not know about the space dimension, but the binary format relies on it.

Another option would be to encode the dimension inside the binary representation.

> Would it make sense to change the internal storage to a 64 bit unsigned int as suggested in the TODO comment of the class? This will increase the speed substantially, but it would limit the maximum number of levels in 3D to 21, and in 2D to 32, is this acceptable? Are there currently limits on the number of levels?

If we only want to support isotropic refinement and meshes supported by p4est, we could pack everything into a 64bit int: you just need a single flag for refinement. Additionally, p4est has a maximum level given by ``dealii::internal::p4est::functions<dim>::max_level`` (30 and 19 I think).

Maybe we should just introduce a second (compact) representation? Inside you can assert that the assumptions are valid.

> Is there a reason CellAccessor<dim,spacedim>::id() is defined in the header include/deal.II/grid/tria_accessor.h?

No, just move it to cellid.h

gassmoeller commented:
Thanks for all the comments. I will think a bit about your suggestions, and try to come up with a design that is both fast enough for my purposes and as compatible as possible with previous behavior (also for the case where DEAL_II_WITH_P4EST is false).

bangerth commented:
/run-tests

gassmoeller commented on source/grid/cell_id.cc:
```diff
@@ -21,13 +21,110 @@
+#ifdef DEAL_II_WITH_CXX11
+std::string
+CellId::to_string() const
+{
```
Yes the result of the function is the same, but the C++11 version is significantly faster (a factor of 2 if I remember). It is a leftover of my first try to speed up the string version though, and I would not object to remove it again, now that there is the binary version.

bangerth replied:
If the two are functionally equivalent, let's keep only one of the two implementations to minimize the effort of maintaining code.

bangerth commented on include/deal.II/distributed/p4est_wrappers.h:
```diff
@@ -272,7 +272,7 @@ namespace internal
       size_t (&connectivity_memory_used) (types<2>::connectivity *p4est);
 
-      static const unsigned max_level;
+      static const unsigned max_level = P4EST_MAXLEVEL;
```
while you're there, can you make it `unsigned int` instead of `unsigned`? Not wrong, just for consistency.

Same in line 457.

bangerth commented on include/deal.II/grid/cell_id.h:
```diff
@@ -48,34 +52,61 @@ template <int, int> class Triangulation;
          const std::vector<unsigned char> &child_indices);
 
   /**
+   * Construct a CellId object with a given @p coarse_cell_index and array of
```
-> coarse_cell_id

bangerth commented on include/deal.II/grid/cell_id.h:
```diff
@@ -48,34 +52,61 @@ template <int, int> class Triangulation;
+   */
+  CellId(const unsigned int coarse_cell_id,
+         const unsigned int n_child_indices,
+         const unsigned char *child_indices);
```
can you say in the docs how `child_indices` is interpreted? I imagine that each byte corresponds to a number between zero and the number of children per cell in whatever the space dimension is this object corresponds to?

bangerth commented on include/deal.II/grid/cell_id.h:
```diff
@@ -48,34 +52,61 @@ template <int, int> class Triangulation;
-  template<int dim, int spacedim>
+  template<int dim, int spacedim=dim>
   typename Triangulation<dim,spacedim>::cell_iterator
   to_cell(const Triangulation<dim,spacedim> &tria) const;
```
the `=dim` in the template arg list requires C++11. But you don't need the default value -- the compiler can deduce both template types.

bangerth commented on source/distributed/p4est_wrappers.cc:
```diff
@@ -191,8 +191,6 @@ namespace internal
     size_t (&functions<2>::connectivity_memory_used) (types<2>::connectivity *p4est)
       = p4est_connectivity_memory_used;
 
-    const unsigned int functions<2>::max_level = P4EST_MAXLEVEL;
```
Even though you put the initializer into the `.h` file, you still need to tell the compiler to allocate a global memory location for this variable (in case, for example, someone wanted to pass around the address of the `static const` variable). So you need to leave this line here, but remove the initializer (which is now at the declaration). Same in line 382.

bangerth commented on source/grid/cell_id.inst.in:
```diff
@@ -21,3 +21,8 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension :  SPACE_DIMENSIONS
+for (deal_II_dimension : DIMENSIONS)
+{
+    template CellId::binary_type CellId::to_binary<deal_II_dimension>() const;
+}
```
endline missing

bangerth commented on tests/grid/cell_id_04.cc:
```diff
@@ -0,0 +1,91 @@
+
+      const CellId cid2(cids);
+
+      typename Triangulation<dim>::cell_iterator cell2 = cid2.to_cell(tr);
```
I suppose you expect `cell==cell2` here, right? If so, maybe also `Assert` that that's so.

gassmoeller commented:
I uploaded a new approach and addressed the comments. The child indices are now stored in a fixed-size array of size MAX_P4EST_LEVEL_2D, or 30 (depending on DEAL_II_WITH_P4EST). The whole cycle of "Given a cell iterator, create a CellId object, serialize it into string/binary, deserialize and create a new cell iterator" now takes 12 times less time (12,000 CPU cycles to 1,000 CPU cycles), and that is fast enough for my purpose. All old use cases should continue to work. There are two limitations of the new approach:
- If we ever get over 30 levels (does anyone see this happening in the next 10 years?) we need to change the internal storage type and the binary type. User code will only be affected by this, when it relies on the fact that binary type is 4 unsigned int's, which would break in that case. Using the typedef I provided should prevent part of the problem (sizeof(binary_type) always gives the correct size in bytes).
- 3D binary_types can only store 20 levels instead of the 30 that can be stored inside of CellID, so there might be a case, where CellId still works, but the binary_type is not large enough. This will throw an exception. A 3D mesh with 10 levels is already huge for my application (billions of unknowns), but I do not know if there applications of deal.II that use many more levels (but again, as soon as they use p4est, they can not be larger than 19 levels anyway).

gassmoeller commented:
@bangerth: You are faster in reviewing than I can describe my changes :smile:. I will address the new comments later today.

bangerth commented:
@gassmoeller -- I know I know :-) One could argue that you're too slow in describing your changes ;-)

Re limitations: No, I'm not worried. As long as our limitations are no stricter than those of p4est, nothing bad can happen. If you're worried, add an assertion and we can tinker with the implementation once someone actually hits the limit.

gassmoeller commented:
Addressed all comments. Ready from my point of view.
